### PR TITLE
[feature] Implementing syncCopyFromURL

### PR DESCRIPTION
**Pull request by AlexanderYW**
Hi, I have implemented syncCopyFromURL

I'm not quite sure about types for it, is there a place i can read about those?, or is the documentation for JS SDK valid in this case?
https://docs.microsoft.com/en-us/javascript/api/@azure/storage-blob/blobclient?view=azure-node-latest#synccopyfromurl-string--blobsynccopyfromurloptions-
If so whats the reason for startCopyFromURL having `tier` but syncCopyFromURL does not? (or aleast I can't find it)

Also prettier ran on `models.ts` on save, replacing ' with " as used in other files :)

### CODE REVIEW

**Comment by AlexanderYW** [comment]
Issue: #461

**Comment by XiaoningLiu** [comment]
Thanks @AlexanderYW for your great contribution! I'm OOF these days will take detailed look at next Monday,

**Comment by AlexanderYW** [comment]
@XiaoningLiu did you get the time to look at my PR :)?

Would be nice if this feature could be a part of Azurite soon, since I want to use it in my own repo to test my driver for AdonisJs :)

File: src/blob/generated/artifacts/models.ts [hunk 1]
```diff
@@ -154,6 +154,7 @@ export interface BlobProperties {
    * Possible values include: 'pending', 'success', 'aborted', 'failed'
    */
   copyStatus?: CopyStatusType;
+  syncCopyStatus?: SyncCopyStatusType;
```
Comment (XiaoningLiu) [inline src/blob/generated/artifacts/models.ts:157]: >syncCopyStatus?: SyncCopyStatusType; [](start = 2, length = 36)

This file (all files under generated folder) should be generated. 

You can extend generated mode in IBlobMetadataStore.ts if generated model doesn't fit our needs.

Comment (XiaoningLiu) [reply src/blob/generated/artifacts/models.ts:157]: Can we leverage existing `copyStatus` member?

---
In reply to: [457872339](https://github.com/Azure/Azurite/pull/469#discussion_r457872339) [](ancestors = 457872339)

File: src/blob/generated/artifacts/models.ts [hunk 2]
```diff
@@ -5013,7 +5014,7 @@ export interface BlobCopyFromURLHeaders {
   /**
    * State of the copy operation identified by x-ms-copy-id. Possible values include: 'success'
    */
-  copyStatus?: SyncCopyStatusType;
+  copyStatus?: CopyStatusType;
```
Comment (XiaoningLiu) [inline src/blob/generated/artifacts/models.ts:5017]: >CopyStatusType [](start = 15, length = 14)

Why this change?

File: src/blob/handlers/BlobHandler.ts [hunk 3]
```diff
@@ -721,6 +713,78 @@ export default class BlobHandler extends BaseHandler implements IBlobHandler {
...
+    options: Models.BlobCopyFromURLOptionalParams,
+    context: Context
+  ): Promise<Models.BlobCopyFromURLResponse> {
+    const blobCtx = new BlobStorageContext(context);
+    const account = blobCtx.account!;
+    const container = blobCtx.container!;
+    const blob = blobCtx.blob!;
+
+    // TODO: Check dest Lease status, and set to available if it's expired, see sample in BlobHandler.setMetadata()
+    const url = new URL(copySource);
```
Comment (XiaoningLiu) [inline src/blob/handlers/BlobHandler.ts:736]: BlobMetadataStore has implemented this in `new BlobWriteLeaseSyncer(doc).sync(lease);` .  Can do same thing in metatadata store implementation.

**Comment by blueww** [comment]
This PR generally looks good to me.
It seems the blob copy will always  finish in the same request. This is true if inside same account, but if cross account copy, the copy blob request might won't finish in the copyblob request, but need check by get blob copy status later. So Azurite behavior might be different than server on copy big blob across account.

File: src/blob/persistence/LokiBlobMetadataStore.ts [hunk 4]
```diff
@@ -1935,6 +1935,181 @@ export default class LokiBlobMetadataStore
...
+    );
+
+    validateWriteConditions(
+      context,
+      options.modifiedAccessConditions,
+      destBlob
+    );
+
+    if (destBlob) {
+      new BlobWriteLeaseValidator(options.leaseAccessConditions).validate(
```
Comment (XiaoningLiu) [inline src/blob/persistence/LokiBlobMetadataStore.ts:2002]: >BlobWriteLeaseValidator [](start = 10, length = 23)

set lease state to available if it's expired by. 

```javascript
lease = new BlobLeaseAdapter(destBlob);
new BlobWriteLeaseSyncer(destBlob).sync(lease);
```

File: src/blob/authentication/OperationBlobSASPermission.ts [hunk 5]
```diff
@@ -420,6 +427,13 @@ OPERATION_BLOB_SAS_CONTAINER_PERMISSIONS.set(
   Operation.Blob_AbortCopyFromURL,
   new OperationBlobSASPermission(BlobSASPermission.Write)
 );
+OPERATION_BLOB_SAS_CONTAINER_PERMISSIONS.set(
+  Operation.Blob_CopyFromURL,
+  // TODO: If destination is an existing blob, create permission is not enough
+  new OperationBlobSASPermission(
```
Comment (XiaoningLiu) [inline src/blob/authentication/OperationBlobSASPermission.ts:433]: We don't need this TODO comment since AccountSASAuthenticator and BlobSASAuthenticator already check this.

**Comment by XiaoningLiu** [comment]
> This PR generally looks good to me.
> It seems the blob copy will always finish in the same request. This is true if inside same account, but if cross account copy, the copy blob request might won't finish in the copyblob request, but need check by get blob copy status later. So Azurite behavior might be different than server on copy big blob across account.

I think current implementation doesn't support cross account blob copy, which is fine for now.

**Comment by XiaoningLiu** [comment]
@AlexanderYW Sorry for late. General looks good with some comments, please take a look.

File: src/blob/generated/artifacts/models.ts (hunk 1 above)
Comment (AlexanderYW) [reply src/blob/generated/artifacts/models.ts:157]: Previously I had issues with building `autorest.typescript.server`, "solution" was to downgrade to nodejs v11.15.0 since the project is using gulp ^3.9.1

Now i only get this type error when generating
image

Beside that it generates the file now

File: src/blob/generated/artifacts/models.ts (hunk 2 above)
Comment (AlexanderYW) [reply src/blob/generated/artifacts/models.ts:5017]: To be honest i don't remember why I did that... but it seems to be working with `SyncCopyStatusType` as well :)

File: src/blob/persistence/LokiBlobMetadataStore.ts (hunk 4 above)
Comment (AlexanderYW) [reply src/blob/persistence/LokiBlobMetadataStore.ts:2002]: Sorry i'm new to the codebase, and I'm not quite sure I understand what you mean

You wan't me to replease the WriteLeaseValidator with your suggestion or?

File: src/blob/authentication/OperationBlobSASPermission.ts (hunk 5 above)
Comment (AlexanderYW) [reply src/blob/authentication/OperationBlobSASPermission.ts:432]: I have removed this TODO in my latest commit, should I remove the one from `Blob_StartCopyFromURL` as well then? :)

Comment (XiaoningLiu) [reply src/blob/authentication/OperationBlobSASPermission.ts:432]: Yes, feel free to remove all of them.

File: src/blob/persistence/LokiBlobMetadataStore.ts (hunk 4 above)
Comment (XiaoningLiu) [reply src/blob/persistence/LokiBlobMetadataStore.ts:2002]: This is for one of your TODO comment in Handler implementation, which mentions need to set lease to avaiable when blob lease is expiry. So the recommendation is to add (instead of removing any) above 2 lines for that.

File: src/blob/generated/artifacts/models.ts (hunk 1 above)
Comment (XiaoningLiu) [reply src/blob/generated/artifacts/models.ts]: You don't need to do a regeneration if this PR doesn't modify any files under generated foler. So I'm thinking if we can keep generated/artifacts/models.ts as it is. Any manual updates to files under generated folder will be override by next code generation.

What's the main purpose for adding `syncCopyStatus` property? If we can reuse `copyStatus` then don't need update the generated/artifacts/models.ts.

Or if it is a must to add `syncCopyStatus`, we can doing this by interface extend (see many other extended models in IBlobMetadataStore.ts), without manually updating generated/artifacts/models.ts. For example

```javascript
interface ExtendBlobProperties extends BlobProperties {
  syncCopyStatus?: SyncCopyStatusType
}
```

**Comment by XiaoningLiu** [comment]
Hi @AlexanderYW, master branch has updated to use Azure Storage track2 SDK v12 for test cases. APIs in v12 is bit different with v10. Please check if any updates needed for test cases in this PR.

**Comment by AlexanderYW** [comment]
Hi @XiaoningLiu, from what i can see there is no merge conflicts with the master branch and my changes, will check if any updates is needed for test cases later today if i get the time :)

File: src/blob/persistence/LokiBlobMetadataStore.ts (hunk 4 above)
Comment (AlexanderYW) [reply src/blob/persistence/LokiBlobMetadataStore.ts:2004]: Hi @XiaoningLiu, I'm afraid I don't dont fully understand what they're doing, but I have tried implementing them

**Comment by XiaoningLiu** [comment]
>     const result_startcopy = await destBlobClient.beginCopyFromURL(

Please keep useing `syncCopyFromURL` .

`beginCopyFromURL` is async copy

---
Refers to: tests/blob/apis/blob.test.ts:786 in c68d8c9. [](commit_id = c68d8c96c33a6ae61f4a4de40923743faf5426c8, deletion_comment = False)

**Comment by AlexanderYW** [comment]
@XiaoningLiu when is this expected to be released :)?

**Comment by XiaoningLiu** [comment]
Hi @AlexanderYW We have a planned release in incoming several weeks according to backend service deployment progress. Promisingly around end of Sep.

**Comment by AlexanderYW** [comment]
Thanks for the update @XiaoningLiu :)
